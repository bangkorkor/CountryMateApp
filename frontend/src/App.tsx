import React, { useEffect, useState } from 'react';
import './App.css';
import { BrowserRouter as Router, Route, Routes, useNavigate, useLocation } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import HomePageComponent from './components/HomePageComponent';
import SettingsPageComponent from './components/SettingsPageComponent';
import FavoritePageComponent from './components/FavoritePageComponent';
import CountriesPageComponent from './components/CountriesPageComponent';

const App: React.FC = () => {
  const [isDarkMode, setIsDarkMode] = useState(false);

  const applyTheme = (mode: 'dark' | 'light') => {
    if (mode === 'dark') {
      document.documentElement.style.setProperty('--primary_color', '#3A4D39'); // Dark primary color
      document.documentElement.style.setProperty('--secondary_color', '#4F6F52'); // Dark secondary color
      document.documentElement.style.setProperty('--tertiary_color', '#ECE3CE'); // Dark secondary color
      document.documentElement.style.setProperty('--quaternary_color', '#ECE3CE'); // Dark secondary color
    } else {
      document.documentElement.style.setProperty('--primary_color', '#ECE3CE'); // Light primary color
      document.documentElement.style.setProperty('--secondary_color', '#739072'); // Light secondary color
      document.documentElement.style.setProperty('--tertiary_color', '#F1EEDC'); // Dark secondary color
      document.documentElement.style.setProperty('--quaternary_color', '#3A4D39'); // Dark secondary color
    }
  };

  // Function to toggle the theme and update localStorage
  const handleThemeToggle = () => {
    setIsDarkMode((prevMode) => {
      const newMode = !prevMode;
      const theme = newMode ? 'dark' : 'light';
      applyTheme(theme);
      localStorage.setItem('theme', theme); // Save the theme preference
      return newMode;
    });
  };

  // Load theme from localStorage on component mount
  useEffect(() => {
    const savedTheme = localStorage.getItem('theme') || 'light'; // Default to light mode
    if (savedTheme === 'dark' || savedTheme === 'light') {
      setIsDarkMode(savedTheme === 'dark');
      applyTheme(savedTheme);
    } else {
      // Default to light mode if the saved theme is invalid or not set
      setIsDarkMode(false);
      applyTheme('light');
    }
  }, []);

  const queryClient = new QueryClient();

  return (
    <QueryClientProvider client={queryClient}>
      <Router basename="/project1">
        {/* RouteHandler stays the same if it's handling other routing logic */}
        <RouteHandler />
        <Routes>
          <Route path="/" element={<HomePageComponent />} />
          <Route path="favorites" element={<FavoritePageComponent />} />
          <Route
            path="settings"
            element={<SettingsPageComponent isDarkMode={isDarkMode} onToggleTheme={handleThemeToggle} />}
          />
          <Route path="continents/:continent" element={<CountriesPageComponent />}></Route>
        </Routes>
      </Router>
    </QueryClientProvider>
  );
};

//This functional component handles the logic for when the user navigates to a new page, and the new page is stored in the localstorage.
const RouteHandler: React.FC = () => {
  const navigate = useNavigate();
  const location = useLocation();

  //This hook sets the "currentRoute" item in localstorage to the page the user is currently at. This updates every time the user navigates to a new page.
  useEffect(() => {
    localStorage.setItem('currentRoute', location.pathname);
  }, [location]);

  //When the page loads, this hook checks if there are any stored routes, and if so, it navigates to that page.
  //If the user is at the page that is stored, it does not do anything.
  useEffect(() => {
    const storedRoute = localStorage.getItem('currentRoute');
    if (storedRoute && storedRoute !== location.pathname) {
      navigate(storedRoute);
    }
  }, [navigate, location.pathname]);

  return null;
};

export default App;
