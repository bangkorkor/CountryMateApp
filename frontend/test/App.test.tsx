import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import App from '../src/App'; 
import { MemoryRouter } from 'react-router-dom';

// Mock localStorage
beforeEach(() => {
  vi.spyOn(Storage.prototype, 'getItem').mockImplementation((key) => {
    if (key === 'theme') return 'light';
    if (key === 'currentRoute') return '/';
    return null;
  });

  vi.spyOn(Storage.prototype, 'setItem').mockImplementation(() => {});
});

describe('App Component', () => {
  it('sets current route to localStorage when navigating', () => {
    // Render the App
    render(<App />);

    // Simulate a route change to /favorites
    window.history.pushState({}, 'Favorites page', '/favorites');

    // Trigger the window's popstate event to ensure the router knows about the route change
    fireEvent.popState(window);

    // Now check if localStorage.setItem was called with the expected arguments
    expect(localStorage.setItem).toHaveBeenCalledWith('currentRoute', '/favorites');
  });
});

  
  it('sets current route to localStorage when navigating', () => {
    render(<App />);
  
    // Simulate a route change to /favorites
    window.history.pushState({}, 'Favorites page', '/favorites');
    fireEvent.popState(window);
  
    // Now check if localStorage.setItem was called with the expected arguments
    expect(localStorage.setItem).toHaveBeenCalledWith('currentRoute', '/favorites');
  });
  

  
  
  
  
  
  
  
  
