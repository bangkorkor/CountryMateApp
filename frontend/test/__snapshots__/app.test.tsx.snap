// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Component > renders correctly (snapshot) 1`] = `
<div>
  <div
    class="Page"
  >
    <div
      class="overlay"
    >
      <div
        class="card"
      >
        <h1
          class="header"
        >
          How to use MoneyMate
        </h1>
        <div
          class="content"
        >
          <img
            alt="Dog 1"
            class="images"
            src="/src/assets/dog1.jpg"
          />
          <div
            class="arrows"
          >
            <button
              class="left-button"
            />
            <h2
              class="pagetext"
            >
               page 
              1
               
            </h2>
            <button
              class="right-button"
            />
          </div>
        </div>
        <button
          class="closebutton"
        >
          Close tutorial
        </button>
      </div>
    </div>
    <div
      class="navbarbox"
    >
      <button
        class="logobutton"
      >
        <h1
          class="logo"
        >
          CUM
        </h1>
      </button>
      <div
        class="menubox"
      >
        <button
          class="navbox"
        >
          <h1
            class="navtext underline"
          >
            Homepage
          </h1>
        </button>
        <button
          class="navbox"
        >
          <h1
            class="navtext "
          >
            Favorites
          </h1>
        </button>
        <button
          class="navbox"
        >
          <h1
            class="navtext "
          >
            Settings
          </h1>
        </button>
      </div>
    </div>
    <div
      class="Text"
    >
      <h1
        class="HomePageTitle"
      >
         Welcome to CountryMate! 
      </h1>
      <h3
        class="HomePageDescription"
      >
         Start by selecting a continent 
      </h3>
      <button
        class="showtutorial"
      >
        Show tutorial again
         
      </button>
    </div>
    <div
      class="ContinentCards"
    >
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          North America
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/north-america1.png"
        />
      </div>
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          South America
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/south-america1.png"
        />
      </div>
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          Africa
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/africa1.png"
        />
      </div>
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          Europe
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/europe1.png"
        />
      </div>
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          Asia
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/asia1.png"
        />
      </div>
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          Oceania
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/australia1.png"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`App Component > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="Page"
  >
    <div
      class="overlay"
    >
      <div
        class="card"
      >
        <h1
          class="header"
        >
          How to use MoneyMate
        </h1>
        <div
          class="content"
        >
          <img
            alt="Dog 1"
            class="images"
            src="/src/assets/dog1.jpg"
          />
          <div
            class="arrows"
          >
            <button
              class="left-button"
            />
            <h2
              class="pagetext"
            >
               page 
              1
               
            </h2>
            <button
              class="right-button"
            />
          </div>
        </div>
        <button
          class="closebutton"
        >
          Close tutorial
        </button>
      </div>
    </div>
    <div
      class="navbarbox"
    >
      <button
        class="logobutton"
      >
        <h1
          class="logo"
        >
          CUM
        </h1>
      </button>
      <div
        class="menubox"
      >
        <button
          class="navbox"
        >
          <h1
            class="navtext underline"
          >
            Homepage
          </h1>
        </button>
        <button
          class="navbox"
        >
          <h1
            class="navtext "
          >
            Favorites
          </h1>
        </button>
        <button
          class="navbox"
        >
          <h1
            class="navtext "
          >
            Settings
          </h1>
        </button>
      </div>
    </div>
    <div
      class="Text"
    >
      <h1
        class="HomePageTitle"
      >
         Welcome to CountryMate! 
      </h1>
      <h3
        class="HomePageDescription"
      >
         Start by selecting a continent 
      </h3>
      <button
        class="showtutorial"
      >
        Show tutorial again
         
      </button>
    </div>
    <div
      class="ContinentCards"
    >
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          North America
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/north-america1.png"
        />
      </div>
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          South America
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/south-america1.png"
        />
      </div>
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          Africa
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/africa1.png"
        />
      </div>
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          Europe
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/europe1.png"
        />
      </div>
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          Asia
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/asia1.png"
        />
      </div>
      <div
        class="ContinentCard"
      >
        <h1
          class="ContinentHeading"
        >
          Oceania
        </h1>
        <img
          alt="Continent"
          class="ContinentPicture"
          src="/src/assets/australia1.png"
        />
      </div>
    </div>
  </div>
</div>
`;
